storage-node/src/cache/data_store_cache/memdisk/cache_manager.rs

```diff
@@ -0,0 +1,361 @@
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl<R: DataStoreReplacer> DataStoreCache for MemDiskStoreCache<R> {
+    async fn get_data_from_cache(
+        &mut self,
+        remote_location: String,
+    ) -> ParpulseResult<Option<Receiver<ParpulseResult<Bytes>>>> {
+        // TODO: Refine the lock.
+        if let Some(mem_store) = &self.mem_store {
+            let mut mem_replacer = self.mem_replacer.as_ref().unwrap().write().await;
+            if let Some((data_store_key, _)) = mem_replacer.get(&remote_location) {
+                match mem_store.read_data(data_store_key) {
+                    Ok(Some(rx)) => return Ok(Some(rx)),
+                    Ok(None) => {
+                        return Err(ParpulseError::Internal(
+                            "Memory replacer and memory store keys are inconsistent.".to_string(),
+                        ))
+                    }
+                    Err(e) => return Err(e),
+                }
+            }
+        }
+
+        let mut disk_replacer = self.disk_replacer.write().await;
+        match disk_replacer.get(&remote_location) {
+            Some((data_store_key, _)) => {
+                if let Some(data) = self.disk_store.read_data(data_store_key).await? {
+                    Ok(Some(data))
+                } else {
+                    unreachable!()
+                }
+            }
+            None => Ok(None),
+        }
+    }
+
+    async fn put_data_to_cache(
+        &mut self,
+        remote_location: String,
+        mut data_stream: StorageReaderStream,
+    ) -> ParpulseResult<usize> {
+        // TODO: Refine the lock.
+        // TODO(lanlou): Also write the data to network.
+        let mut bytes_to_disk = None;
+        let mut bytes_mem_written = 0;
+        let mut evicted_bytes_to_disk: Option<
+            Vec<(ParpulseDataStoreCacheKey, (Vec<Bytes>, usize))>,
+        > = None;
+        // 1. If the mem_store is enabled, first try to write the data to memory.
+        // Note: Only file which size < mem_max_file_size can be written to memory.
+        if let Some(mem_store) = &mut self.mem_store {
+            loop {
+                match data_stream.next().await {
+                    Some(Ok(bytes)) => {
+                        bytes_mem_written += bytes.len();
+                        // Need to write the data to network.
+                        if let Some((bytes_vec, _)) =
+                            mem_store.write_data(remote_location.clone(), bytes)
+                        {
+                            // If write_data returns something, it means the file size is too large
+                            // to fit in the memory. We should put it to disk cache.
+                            bytes_to_disk = Some(bytes_vec);
+                            break;
+                        }
+                    }
+                    Some(Err(e)) => return Err(e),
+                    None => break,
+                }
+            }
```

Comment:
Shall we check the max capacity of `mem_replacer` here instead of directly inserting into the mem store and then checking whether the size is valid according to the `mem_replacer`?

Reply:
I put a check in the new method per discussion.